add examples that demonstrate how to use mergeItem, multiMerge, multi…

geirman opened the pull request:
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request:

(You can skip this if you're fixing a typo or adding an app to the Showcase.)

Explain the **motivation** for making this change. What existing problem does the pull request solve?

it wasn't clear how to use various methods such as mergeItem, multiMerge, etc, so after figuring it out I thought it would be useful to provide clear examples for others. 

Example: When "Adding a function to do X", explain why it is necessary to have a way to do X.

**Test plan (required)**

Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI.

Make sure tests pass on both Travis and Circle CI.

**Code formatting**

See the simple [style guide](https://github.com/facebook/react-native/blob/master/CONTRIBUTING.md#style-guide).

…Set, multiGet, and multiRemove

vjeux commented:
Awesome, could you attach a screenshot of how it looks like on the website?

geirman commented:
@vjeux I'm not sure what I'd screenshot. All the outputs are console.logs. You want screenshots of those? Also, I see eslint-bot is complaining of trailing spaces. Should I remove those and resubmit, or do those get stripped away automagically?

geirman commented:
@vjeux I get what you mean by screenshot now. How would I render the output so that I can capture a screenshot?

rmcvey commented:
Thanks for adding the examples, they look good. Regarding the whitespace issues, I see trailing whitespace on the following lines:

108, 109, 161, 220, 271, 293, 322, 327, 331, 334

There is also a minor formatting error on 339 `=>{` should be `=> {`

vjeux commented:
If you go to the website folder and do npm install and npm start, it'll open a local copy of the website and you can see how the async storage docs look like.

geirman commented:
Thanks for that tip @vjeux!  As it turns out, these edits DO NOT look great. I tried wrapping in `&lt;pre&gt;` tags as well with no luck. Any suggestions?

Here's what I'm seeing...
![image](https://cloud.githubusercontent.com/assets/1640318/13725409/c81e9766-e855-11e5-8562-65511beb6296.png)

Also, would you welcome a PR to the Contributing document to describe how to build and run the website? https://github.com/facebook/react-native/blob/master/CONTRIBUTING.md

vjeux commented:
You can do

```
 ```javascript
 Some_code()
 ```
```

in the comment to get syntax highlighting

geirman commented:
@rmcvey @vjeux thanks for reviewing and for your patience. I've fixed the formatting concerns and verified that there is some precedence for providing code examples such as these. I've also squashed the commits. Please let me know if there's anything more needed. I'm starting to figure this PR thing out.

![image](https://cloud.githubusercontent.com/assets/1640318/13731289/8bda1500-e922-11e5-8176-bc7816a17e2e.png)

vjeux commented:
Yay, now it starts looking good :)

vjeux commented on Libraries/Storage/AsyncStorage.js:
````diff
@@ -105,6 +105,22 @@ var AsyncStorage = {
+   *
+   * Example:
+   * ```JavaScript
+   *   let k1 = 'key_1'
````
Remove all the indent, it shows up on the docs otherwise

vjeux commented on Libraries/Storage/AsyncStorage.js:
````diff
@@ -105,6 +105,22 @@ var AsyncStorage = {
+   * Example:
+   * ```JavaScript
+   *   let k1 = 'key_1'
+   *   let val_1a = JSON.stringify({'a': '123', 'b': {'b1': '234', 'b2': '234'}})
````
Please add ; at the end of the lines

vjeux commented on Libraries/Storage/AsyncStorage.js:
````diff
@@ -105,6 +105,22 @@ var AsyncStorage = {
    * Returns a `Promise` object. Not supported by all native implementations.
+   *
+   * Example:
+   * ```JavaScript
````
lower case javascript to be compatible with github (the current implementation only supports js so it throws it away fyi)

vjeux commented on Libraries/Storage/AsyncStorage.js:
````diff
@@ -201,6 +219,17 @@ var AsyncStorage = {
+   * ```JavaScript
+   *   AsyncStorage.getAllKeys( (err, keys) => {
+   *     AsyncStorage.multiGet( keys, (err, stores) => {
+   *      stores.map( (result, i, store) => {
````
Please no space around )

vjeux replied:
(

vjeux commented on Libraries/Storage/AsyncStorage.js:
````diff
@@ -263,6 +294,14 @@ var AsyncStorage = {
+   *
+   * Example:
+   * ```JavaScript
+   *   let keys = ['k1','k2']
````
Space after `,`

vjeux commented on Libraries/Storage/AsyncStorage.js:
```diff
@@ -286,6 +325,32 @@ var AsyncStorage = {
+   *         })
+   *       })
+   *     })
+   *   })
```
Maybe I'm missing something but this example doesn't make it clearer to me what multimerge does.

geirman replied:
Ok, I'll think on this more

vjeux commented on Libraries/Storage/AsyncStorage.js:
```diff
@@ -105,6 +105,22 @@ var AsyncStorage = {
+   *   let k1 = 'key_1'
+   *   let val_1a = JSON.stringify({'a': '123', 'b': {'b1': '234', 'b2': '234'}})
+   *   let val_1b = JSON.stringify({'b': {'b2': '555', 'b3': '555'}})
+   *   let EXPECTED_RESULT = {'a':'123','b':{'b1':'234','b2':'555','b3':'555'}}
```
Instead of a/b and 123/234/555 could you use some more realistic names and values. Otherwise it is pretty hard to figure out what is going on. I have to spend a lot of time read through the two json dumps to figure things out.

geirman replied:
sure, I'll give it more thought

vjeux commented on Libraries/Storage/AsyncStorage.js:
````diff
@@ -105,6 +105,22 @@ var AsyncStorage = {
+   * ```JavaScript
+   *   let k1 = 'key_1'
+   *   let val_1a = JSON.stringify({'a': '123', 'b': {'b1': '234', 'b2': '234'}})
+   *   let val_1b = JSON.stringify({'b': {'b2': '555', 'b3': '555'}})
````
Do you really need to manually json stringify the object?

geirman replied:
I do, unless you know something I don't. I stringify there to avoid having to do it below, and I AsyncStorage only stores stringified values, else it errors. Or am I missing something?

geirman commented:
@vjeux @rmcvey I've updated the examples to clear up the formatting concerns you noted. As well, I tried to make the multiMerge & mergeItem examples more realistic. Let me know what you think. Will squash once I get the thumbs up.

vjeux commented:
Awesome, looks really good now :)

It makes me want to have this kind of examples all over our docs now :p

You don't need to squash, when we import it is turned into a single commit.

@facebook-github-bot shipit

geirman commented:
@vjeux Awesome! Thanks for coaching me. Now that I know what needs doing, I'll work on others as well, starting with highlightRow in [ListView.renderRow](http://facebook.github.io/react-native/docs/listview.html#renderrow). That really needs an example. After a couple hours, I was unable to figure it out myself. I'll keep trying, but if you have any advice or pointers, please clue me in!

rmcvey commented:
@geirman If you are still having trouble finding it, trailing whitespace are on lines 114, 122, 346, 359, 367, and 370

geirman commented:
@rmcvey thanks! I've now installed a plugin to automatically trim trailing whitespace for me, so it shouldn't be a problem going forward.

rmcvey commented:
@geirman good call, better when you don't have to think about it at all :)

vjeux commented:
@facebook-github-bot shipit

vjeux commented on Libraries/Storage/AsyncStorage.js:
```diff
@@ -105,6 +105,30 @@ var AsyncStorage = {
+   *  age: 30,
+   *  traits: {hair: 'brown', eyes: 'brown'},
+   * };
+
```
please add * in all the lines, even empty ones

vjeux commented on Libraries/Storage/AsyncStorage.js:
```diff
@@ -105,6 +105,30 @@ var AsyncStorage = {
+   *  traits: {hair: 'brown', eyes: 'brown'},
+   * };
+
+   // need only define what will be added or updated
```
missing *

vjeux commented on Libraries/Storage/AsyncStorage.js:
```diff
@@ -105,6 +105,30 @@ var AsyncStorage = {
+   *   AsyncStorage.mergeItem('UID123', JSON.stringify(UID123_delta), () => {
+   *     AsyncStorage.getItem('UID123', (err, result) => {
+   *       console.log(result);
+   *       // => {'name':'Chris','age':31,'traits':{'shoe_size':10,'hair':'brown','eyes':'blue'}}
```
missing space after `:`, also please remove quotes around keys

vjeux commented on Libraries/Storage/AsyncStorage.js:
```diff
@@ -105,6 +105,30 @@ var AsyncStorage = {
+   // need only define what will be added or updated
+   * let UID123_delta = {
+   *  age: 31,
+   *  traits: {eyes: 'blue', shoe_size: 10}
```
in js, we mostly use shoeSize capitalizaiton

vjeux commented on Libraries/Storage/AsyncStorage.js:
```diff
@@ -286,6 +336,52 @@ var AsyncStorage = {
+   *         let key = store[i][0];
+   *         let val = store[i][1];
+   *         console.log(key, val);
+   *         // => UID234 {"name":"Chris","age":31,"traits":{"shoe_size":10,"hair":"brown","eyes":"blue"}}
```
use single quote

geirman commented:
@vjeux thanks again, so sorry for all the nit noid issues. Hope we're good to go now.

vjeux commented:
@facebook-github-bot shipit
